manager/secret_manager.go

```diff
@@ -1,13 +1,114 @@
 package manager
 
 import (
+	"encoding/base64"
+	"encoding/json"
+	goerrors "errors"
 	"github.com/onepanelio/core/model"
+	"github.com/onepanelio/core/util"
+	"google.golang.org/grpc/codes"
+	"k8s.io/apimachinery/pkg/api/errors"
 )
 
 func (r *ResourceManager) CreateSecret(namespace string, secret *model.Secret) (err error) {
-	return r.kubeClient.CreateSecret(namespace, secret)
+	if err = r.kubeClient.CreateSecret(namespace, secret); err != nil {
+		return util.NewUserError(codes.Unknown, "Secret was not created.")
+	}
+	return
+}
+
+func (r *ResourceManager) SecretExists(namespace string, name string) (exists bool, err error) {
+	var foundSecret *model.Secret
+	var statusError *errors.StatusError
+	secret := model.Secret{
+		Name: name,
+	}
+	foundSecret, err = r.kubeClient.SecretExists(namespace, &secret)
+	if err != nil {
+		if goerrors.As(err, &statusError) {
+			if statusError.ErrStatus.Reason == "NotFound" {
+				return false, util.NewUserError(codes.NotFound, "Secret Not Found.")
+			}
+			return false, util.NewUserError(codes.Unknown, "Error when checking existence of secret.")
+		}
+		return false, util.NewUserError(codes.Unknown, "Error when checking existence of secret.")
+	}
+	if foundSecret.Name == "" {
+		return false, nil
+	}
+	return true, nil
 }
 
 func (r *ResourceManager) GetSecret(namespace, name string) (secret *model.Secret, err error) {
 	return r.kubeClient.GetSecret(namespace, name)
 }
+
+func (r *ResourceManager) ListSecrets(namespace string) (secrets []*model.Secret, err error) {
+	secrets, err = r.kubeClient.ListSecrets(namespace)
+	if err != nil {
+		return nil, util.NewUserError(codes.NotFound, "No secrets were found.")
+	}
+
+	return
+}
+
+func (r *ResourceManager) DeleteSecret(namespace string, name string) (deleted bool, err error) {
+	return r.kubeClient.DeleteSecret(namespace, name)
+}
+
+func (r *ResourceManager) DeleteSecretKey(namespace string, name string, key string) (deleted bool, err error) {
+	return r.kubeClient.DeleteSecretKey(namespace, name, key)
+}
+
+func (r *ResourceManager) AddSecretKeyValue(namespace string, name string, key string, value string) (inserted bool, err error) {
+	return r.kubeClient.AddSecretKeyValue(namespace, name, key, value)
+}
+
+func (r *ResourceManager) UpdateSecretKeyValue(namespace string, secret *model.Secret) (updated bool, err error) {
```

Comment:
@rushtehrani Should the function param here be a model.Secret? Doesn't this mean that the Server is creating the model, instead of passing the data along?
Shouldn't the manager be creating the model?

Reply:
@aleksandrmelnikov no, this is correct, the Server is responsible for converting the api request message to a model and pass into manager.

Reply:
Hmm, okay. So it's okay if the manager takes a model, passes a model, returns a model?

Reply:
Yes